tools/cell_census_builder/validate.py

```diff
@@ -276,59 +276,86 @@ def _validate_X_layers_contents_by_dataset(args: Tuple[str, str, Dataset, List[E
+                exp.obs.read(
+                    column_names=["soma_joinid", "dataset_id"], value_filter=f"dataset_id == '{dataset.dataset_id}'"
                 )
-                assert (
-                    expected_ad_x_coo != actual_x_coo
-                ).nnz == 0, f"{eb.name}:{dataset.dataset_id} the X matrix elements are not equal."
-            expected_gene_count = np.count_nonzero(expected_ad_x.sum(axis=0))
-            assert expected_gene_count == actual_census_gene_count, error.format(
-                "gene_count", actual_census_gene_count, expected_gene_count
+                .concat()
+                .to_pandas()
+            )
+
+            assert ad.n_obs == len(obs_joinids)
+            if len(obs_joinids) == 0:
+                continue
+
+            X_raw = (
+                exp.ms[MEASUREMENT_RNA_NAME]
+                .X["raw"]
+                .read((obs_joinids.soma_joinid.to_numpy(), slice(None)))
+                .tables()
+                .concat()
+            )
+            data = X_raw["soma_data"].to_numpy()
+            rows = X_raw["soma_dim_0"].to_numpy()
+            cols = X_raw["soma_dim_1"].to_numpy()
+
+            # positionally (re)index obs/rows. We _know_ that the Census assigns
+            # obs soma_joinids in the obs position order of the original AnnData, so
+            # leverage that for simplicity and speed.
+            rows_by_position = obs_joinids.set_index("soma_joinid").index.get_indexer(rows)
+
+            # get the joinids for the var axis
+            all_var_ids = (
+                exp.ms[MEASUREMENT_RNA_NAME].var.read(column_names=["soma_joinid", "feature_id"]).concat().to_pandas()
             )
+            # mask defines which feature_ids are in the AnnData
+            var_joinid_in_adata = all_var_ids.feature_id.isin(expected_ad_var.index)
+            assert ad.n_vars == var_joinid_in_adata.sum()
+
+            # positionally re-index the vars/cols using the feature_id as the join key.
+            cols_by_position = (
+                expected_ad_var.join(all_var_ids.set_index("feature_id"))
+                .set_index("soma_joinid")
+                .index.get_indexer(cols)
+            )
+
+            # Assertion 1 - the contents of the X matrix are EQUAL for all var values
+            # present in the AnnData
+            assert (
+                sparse.coo_matrix((data, (rows_by_position, cols_by_position)), shape=expected_ad_x.shape)
+                != expected_ad_x
+            ).nnz == 0, f"{eb.name}:{dataset.dataset_id} the X matrix elements are not equal."
+
+            # Assertion 2 - the contents of the X matrix are EMPTY for all var ids
+            # NOT present in the AnnData. Test by asserting that no col IDs contain
+            # a joinid not in the AnnData.
+            assert (
+                var_joinid_in_adata.all()
+                or not pd.Series(cols).isin(all_var_ids[~var_joinid_in_adata].soma_joinid).any()
```

Comment:
Can you explain this check for me and why we have an `or`?

Reply:
consider the case where _all_ of the values in the mask are True.  In that case, the second clause will fail (the arg to `isin` will be an empty set), so it is protected with the first clause.